File: UnityClient/PlanetServer/Src/Data/PsArray.cs
using System;
using System.Collections.Generic;

namespace PS.Data
{
    /// <summary>
    /// Holds an array of data.
    /// </summary>
    public class PsArray
    {
        /// <summary>
        /// Data for the array.
        /// </summary>
        public List<PsDataWrapper> _content;

        /// <summary>
        /// Create an PsArray from an array of objects.
        /// </summary>
        /// <param name="obj">Array of objects.</param>
        /// <returns></returns>
        public static PsArray Create(object[] obj)
        {
            PsArray psa = new PsArray();

            foreach (object o in obj)
            {
                Dictionary<string, object> dict = (Dictionary<string, object>)o;

                int type = (int)dict["t"];

                switch (type)
                {
                    case (int)Constants.PsType.Boolean:
                        psa.AddBoolean(Convert.ToBoolean(dict["v"]));
                        break;

                    case (int)Constants.PsType.String:
                        psa.AddString(Convert.ToString(dict["v"]));
                        break;

                    case (int)Constants.PsType.Integer:
                        psa.AddInt(Convert.ToInt32(dict["v"]));
                        break;

                    case (int)Constants.PsType.Long:
                        psa.AddLong(Convert.ToInt64(dict["v"]));
                        break;

                    case (int)Constants.PsType.Float:
                        psa.AddFloat(Convert.ToSingle(dict["v"]));
                        break;

                    case (int)Constants.PsType.PSObject:
                        psa.AddPsObject(PsObject.Create((Dictionary<string, object>)dict["v"]));
                        break;

                    case (int)Constants.PsType.PSArray:
                        psa.AddPsArray(PsArray.Create((object[])dict["v"]));
                        break;

                    case (int)Constants.PsType.Number:
                        psa.AddNumber(Convert.ToSingle(dict["v"]));
                        break;
                }
            }

            return psa;
        }

        /// <summary>
        /// Initializes a new instance of the PsArray class.
        /// </summary>
        public PsArray()
        {
            _content = new List<PsDataWrapper>();
        }

        /// <summary>
        /// Look for a boolean value in the array.
        /// </summary>
        /// <param name="key">Key to lookup.</param>
        /// <returns>Boolean value of the key.</returns>
        public bool GetBoolean(int key)
        {
            return (bool)_content[key].v;
        }

        /// <summary>
        /// Look for a string value in the array.
        /// </summary>
        /// <param name="key">Key to lookup.</param>
        /// <returns>String value of the key.</returns>
        public string GetString(int key)
        {
            return (string)_content[key].v;
        }

        /// <summary>
        /// Look for a int value in the array.
        /// </summary>
        /// <param name="key">Key to lookup.</param>
        /// <returns>Int value of the key.</returns>
        public int GetInt(int key)
        {
            return (int)_content[key].v;
        }

        /// <summary>
        /// Look for a long value in the array.
        /// </summary>
        /// <param name="key">Key to lookup.</param>
        /// <returns>Long value of the key.</returns>
        public long GetLong(int key)
        {
            return (long)_content[key].v;
        }

        /// <summary>
        /// Look for a float value in the array.
        /// </summary>
        /// <param name="key">Key to lookup.</param>
        /// <returns>Float value of the key.</returns>
        public float GetFloat(int key)
        {
            return (float)_content[key].v;
        }

        /// <summary>
        /// Look for a double value in the array.
        /// </summary>
        /// <param name="key">Key to lookup.</param>
        /// <returns>Double value of the key.</returns>
        public double GetDouble(int key)
        {
            return (double)_content[key].v;
        }

        /// <summary>
        /// Look for a PsObject value in the array.
        /// </summary>
        /// <param name="key">Key to lookup.</param>
        /// <returns>PsObject value of the key.</returns>
        public PsObject GetPsObject(int key)
        {
            return (PsObject)_content[key].v;
        }

        /// <summary>
        /// Look for a PsArray value in the array.
        /// </summary>
        /// <param name="key">Key to lookup.</param>
        /// <returns>PsArray value of the key.</returns>
        public PsArray GetPsArray(int key)
        {
            return (PsArray)_content[key].v;
        }

        /// <summary>
        /// Look for a (Flash)number value in the array.
        /// </summary>
        /// <param name="key">Key to lookup.</param>
        /// <returns>Number value of the key.</returns>
        public double GetNumber(int key)
        {
            return (double)_content[key].v;
        }

        /// <summary>
        /// Add a boolean to the array.
        /// </summary>
        /// <param name="value">Boolean to add.</param>
        public void AddBoolean(bool value)
        {
            _content.Add(new PsDataWrapper((int)Constants.PsType.Boolean, value));
        }

        /// <summary>
        /// Add a string to the array.
        /// </summary>
        /// <param name="value">String to add.</param>
        public void AddString(string value)
        {
            _content.Add(new PsDataWrapper((int)Constants.PsType.String, value));
        }

        /// <summary>
        /// Add a int to the array.
        /// </summary>
        /// <param name="value">Int to add.</param>
        public void AddInt(int value)
        {
            _content.Add(new PsDataWrapper((int)Constants.PsType.Integer, value));
        }

        /// <summary>
        /// Add a long to the array.
        /// </summary>
        /// <param name="value">Long to add.</param>
        public void AddLong(long value)
        {
            _content.Add(new PsDataWrapper((int)Constants.PsType.Long, value));
        }

        /// <summary>
        /// Add a float to the array.
        /// </summary>
        /// <param name="value">Float to add.</param>
        public void AddFloat(float value)
        {
            _content.Add(new PsDataWrapper((int)Constants.PsType.Float, value));
        }

        /// <summary>
        /// Add a double to the array.
        /// </summary>
        /// <param name="value">Double to add.</param>
        public void AddDouble(double value)
        {
            _content.Add(new PsDataWrapper((int)Constants.PsType.Double, value));
        }

        /// <summary>
        /// Add a PsObject to the array.
        /// </summary>
        /// <param name="value">PsObject to add.</param>
        public void AddPsObject(PsObject value)
        {
            _content.Add(new PsDataWrapper((int)Constants.PsType.PSObject, value));
        }

        /// <summary>
        /// Add a PsArray to the array.
        /// </summary>
        /// <param name="value">PsArray to add.</param>
        public void AddPsArray(PsArray value)
        {
            _content.Add(new PsDataWrapper((int)Constants.PsType.PSArray, value));
        }

        /// <summary>
        /// Add a (Flash)number to the array.
        /// </summary>
        /// <param name="value">Number to add.</param>
        public void AddNumber(double value)
        {
            _content.Add(new PsDataWrapper((int)Constants.PsType.Number, value));
        }

        /// <summary>
        /// Number of items in the array.
        /// </summary>
        public int Count { get { return _content.Count; } }

        /// <summary>
        /// Creates a List of objects from the array.
        /// </summary>
        /// <returns>List of objects.</returns>
        public List<object> ToArrayObject()
        {
            List<object> ret = new List<object>();

            for (int i = 0; i < _content.Count; ++i)
            {
                PsDataWrapper data = _content[i];

                if (data.v is PsObject)
                {
                    ret.Add((data.v as PsObject).ToObject());
                }
                else if (data.v is PsArray)
                {
                    ret.Add((data.v as PsArray).ToArrayObject());
                }
                else
                    ret.Add(data);
            }

            return ret;
        }

        /// <summary>
        /// Creates a string from the data.
        /// </summary>
        /// <returns>String</returns>
        public override string ToString()
        {
            string str = "Array [";
            for (int i = 0; i < Count; ++i)
            {
                str += _content[i].v;
                if (i < Count - 1)
                    str += ",";
            }
            str += "]";

            return str;
        }
    }
}

File: PlanetServer/src/planetserver/network/PsArray.java
package planetserver.network;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

import net.sf.json.JSONArray;
import net.sf.json.JSONObject;

public class PsArray 
{
    public LinkedList<PsDataWrapper> content;

    public PsArray()
    {
        this.content = new LinkedList<PsDataWrapper>();
    }

    public int getInteger(int index)
    {
        return (Integer)this.content.get(index).getV();
    }

    public long getLong(int index)
    {
        return (Long)this.content.get(index).getV();
    }

    public String getString(int index)
    {
        return (String)this.content.get(index).getV();
    }

    public boolean getBoolean(int index)
    {
        int b = (Integer)this.content.get(index).getV();
        return (b == 1);
    }

    public float getFloat(int index)
    {
        return (Float)this.content.get(index).getV();
    }

    public double getDouble(int index)
    {
        return (Double)this.content.get(index).getV();
    }
    
    public PsObject getPsObject(int index)
    {
        return (PsObject)this.content.get(index).getV();
    }
    
    public PsArray getPsArray(int index)
    {
        return (PsArray)this.content.get(index).getV();
    }

    //number from as3 returns double in java
    public double getNumber(int index)
    {
        return (Double)this.content.get(index).getV();
    }

    public void addBoolean(boolean value)
    {
        this.content.add(new PsDataWrapper(value, PsType.TYPE_BOOLEAN));
    }

    public void addString(String value)
    {
        this.content.add(new PsDataWrapper(value, PsType.TYPE_STRING));
    }

    public void addInteger(int value)
    {
        this.content.add(new PsDataWrapper(value, PsType.TYPE_INTEGER));
    }

    public void addLong(Long value)
    {
        this.content.add(new PsDataWrapper(value, PsType.TYPE_LONG));
    }

    public void addFloat(float value)
    {
        this.content.add(new PsDataWrapper(value, PsType.TYPE_FLOAT));
    }

    public void addDouble(double value)
    {
        this.content.add(new PsDataWrapper(value, PsType.TYPE_DOUBLE));
    }

    public void addPsObject(PsObject value)
    {
        this.content.add(new PsDataWrapper(value, PsType.TYPE_PSOBJECT));
    }
    
    public void addPsArray(PsArray value)
    {
        this.content.add(new PsDataWrapper(value, PsType.TYPE_PSARRAY));
    }
    
    //add number as double
    public void addNumber(double value)
    {
        this.content.add(new PsDataWrapper(value, PsType.TYPE_NUMBER));
    }

    public JSONArray toJSONObject()
    {
        JSONArray jArray = new JSONArray();

        for (PsDataWrapper obj : this.content)
        {
            if (obj.getV() instanceof PsObject)
            {
                JSONObject psObject = new JSONObject();
                psObject.put("t", PsType.TYPE_PSOBJECT);
                psObject.put("v", ((PsObject) obj.getV()).toJSONObject());
                jArray.add(psObject);
            }
            else if (obj.getV() instanceof PsArray)
            {
                JSONObject psArray = new JSONObject();
                psArray.put("t", PsType.TYPE_PSARRAY);
                psArray.put("v", ((PsArray) obj.getV()).toJSONObject());
                jArray.add(psArray);
            }
            else
            {
                jArray.add(obj);
            }
        }
        return jArray;
    }

    /**
     * Takes in json object and applies properties to the ps object
     * @param jsonObject 
     */
    public void fromJsonObject(JSONArray jsonObject)
    {
        for (Object jsonValues : jsonObject)
        {
            JSONObject keyObj = (JSONObject) jsonValues;
            int type = keyObj.getInt(("t"));
            Object value = keyObj.get("v");

            switch (type)
            {
                case PsType.TYPE_BOOLEAN:
                    addBoolean((Boolean) value);
                    break;

                case PsType.TYPE_STRING:
                    addString((String) value);
                    break;

                case PsType.TYPE_DOUBLE:
                    addDouble((Double) value);
                    break;

                case PsType.TYPE_FLOAT:
                    addFloat((Float) value);
                    break;

                case PsType.TYPE_INTEGER:
                    addInteger((Integer) value);
                    break;

                case PsType.TYPE_LONG:
                    addLong((Long) value);
                    break;

                case PsType.TYPE_NUMBER:
                    addDouble((Double) value); //just convert number to decimal
                    break;

                case PsType.TYPE_PSARRAY:
                    PsArray newArray = new PsArray();
                    newArray.fromJsonObject((JSONArray) value);
                    addPsArray(newArray);
                    break;

                case PsType.TYPE_PSOBJECT:
                    PsObject newObject = new PsObject();
                    newObject.fromJsonObject((JSONObject) value);
                    addPsObject(newObject);
                    break;

                default:
                    throw new Error("Unsupported Type!");
            }
        }
    }

    ///////////to list functions
    public List<Object> toObjectList()
    {
        List<Object> list = new ArrayList<Object>();

        for (PsDataWrapper data : this.content)
        {
            Object obj = data.getV();

            if (obj instanceof PsObject)
            {
                list.add(((PsObject) obj).toMap());
            }
            else if (obj instanceof PsArray)
            {
                list.add(((PsArray) obj).toObjectList());
            }
            else
            {
                list.add(obj);
            }
        }
        return list;
    }
}
